src/SignalR/server/Core/src/Internal/HubFilterFactory.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    internal class HubFilterFactory<T> : IHubFilter where T : IHubFilter
+    {
+        private readonly ObjectFactory _objectFactory;
+
+        public HubFilterFactory()
+        {
+            _objectFactory = ActivatorUtilities.CreateFactory(typeof(T), Array.Empty<Type>());
+        }
+
+        public ValueTask<object> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
+        {
+            return GetFilter(_objectFactory, invocationContext.ServiceProvider).InvokeMethodAsync(invocationContext, next);
+        }
+
+        public Task OnConnectedAsync(HubInvocationContext context, Func<HubInvocationContext, Task> next)
+        {
+            return GetFilter(_objectFactory, context.ServiceProvider).OnConnectedAsync(context, next);
+        }
+
+        public Task OnDisconnectedAsync(HubInvocationContext context, Exception exception, Func<HubInvocationContext, Exception, Task> next)
+        {
+            return GetFilter(_objectFactory, context.ServiceProvider).OnDisconnectedAsync(context, exception, next);
```

Comment:
My memory of SignalR DI scope and object lifetimes is hazy. How does it work?

I would find it surprising that a new filter instance is created for each event. I would have expected a filter to be created once for a hub connection and the same filter instance to handle each event. That would allow me to potentially put state on the filter.

If that isn't realistic or desirable then that needs to be called out in code comments and documentation.

Reply:
Hubs are created per call so the most logical default is to follow that (for non instance filters).  W e could keep the singleton but it makes it a bit ugly when you want to inject scoped things.